sklearn/utils/__init__.py

```diff
@@ -180,199 +181,166 @@ def axis0_safe_slice(X, mask, len_mask):
-                - string (key-based): only supported for dataframes. No keys
-                  other than strings are allowed.
-    axis : int, default=0
-        The axis along which `X` will be subsampled. ``axis=0`` will select
-        rows while ``axis=1`` will select columns.
+    key : scalar, slice or array-like
+        The key from which we want to infer the data type.
 
     Returns
     -------
-    subset
-        Subset of X on axis 0 or 1.
-
-    Notes
-    -----
-    CSR, CSC, and LIL sparse matrices are supported. COO sparse matrices are
-    not supported.
+    dtype : {'int', 'str', 'bool'}
+        Returns the data type of key.
     """
-    if axis == 0:
-        return _safe_indexing_row(X, indices)
-    elif axis == 1:
-        return _safe_indexing_column(X, indices)
-    else:
-        raise ValueError(
-            "'axis' should be either 0 (to index rows) or 1 (to index "
-            " column). Got {} instead.".format(axis)
-        )
-
+    err_msg = ("No valid specification of the columns. Only a scalar, list or "
+               "slice of all integers or all strings, or boolean mask is "
+               "allowed")
 
-def _array_indexing(array, key, axis=0):
-    """Index an array consistently across NumPy version."""
-    if axis not in (0, 1):
-        raise ValueError(
-            "'axis' should be either 0 (to index rows) or 1 (to index "
-            " column). Got {} instead.".format(axis)
-        )
-    if np_version < (1, 12) or issparse(array):
-        # check if we have an boolean array-likes to make the proper indexing
-        key_array = np.asarray(key)
-        if np.issubdtype(key_array.dtype, np.bool_):
-            key = key_array
-    return array[key] if axis == 0 else array[:, key]
+    dtype_to_str = {int: 'int', str: 'str', bool: 'bool', np.bool_: 'bool'}
+    array_dtype_to_str = {'i': 'int', 'u': 'int', 'b': 'bool', 'O': 'str',
+                          'U': 'str', 'S': 'str'}
 
+    if key is None:
+        return None
+    if isinstance(key, tuple(dtype_to_str.keys())):
+        try:
+            return dtype_to_str[type(key)]
+        except KeyError:
+            raise ValueError(err_msg)
+    if isinstance(key, slice):
+        if key.start is None and key.stop is None:
+            return None
+        key_start_type = _determine_key_type(key.start)
+        key_stop_type = _determine_key_type(key.stop)
+        if key_start_type is not None and key_stop_type is not None:
+            if key_start_type != key_stop_type:
+                raise ValueError(err_msg)
+        if key_start_type is not None:
+            return key_start_type
+        return key_stop_type
+    if isinstance(key, list):
+        unique_key = set(key)
+        set_type = {_determine_key_type(elt) for elt in unique_key}
+        if not set_type:
+            return None
+        if len(set_type) != 1:
+            raise ValueError(err_msg)
+        set_type, = set_type
+        return set_type
```

Comment:
Given `len(set_type)` must be 1 here, does this work?

```
return set_type[0]
```

Reply:
not because set are cannot be indexed. True that I could do

```python
return list(set_type)[0]
```

which is maybe less weird than the unpacking

Reply:
How about

```py
return set_type.pop()
```